### ISSUE TITLE
Backup Cluster Failed with persistentvolumeclaim not found error in server pod log

### ISSUE THREAD

**Issue by smm190** [heptio/ark#450]
I am trying to backup my entire cluster.  The backup is failing.  In the ark server pod log, I am seeing these errors saying that the persistentvolumeclaim is not found:
```
time="2018-04-23T19:34:40Z" level=info msg="Backup completed" backup=heptio-ark/qa-20180423-3 logSource="pkg/controller/backup_controller.go:375"
time="2018-04-23T19:34:41Z" level=error msg="backup failed" error="[persistentvolumeclaims \"brigade-worker-01c9sm9g47trhpg0fmhwb53hc4\" not found, persistentvolumeclaims \"brigade-worker-01ca6455q4taxqdkm23gg27hgh\" not found, persistentvolumeclaims \"brigade-worker-01ca64ywwfnjb5mfx3c1pjyxye\" not found, persistentvolumeclaims \"brigade-worker-01c9sm9g47trhpg0fmhwb53hc4\" not found, persistentvolumeclaims \"brigade-worker-01ca6455q4taxqdkm23gg27hgh\" not found, persistentvolumeclaims \"brigade-worker-01ca64ywwfnjb5mfx3c1pjyxye\" not found, persistentvolumeclaims \"brigade-worker-01canstq5nyaae8a760mb1k37s\" not found, persistentvolumeclaims \"brigade-worker-01c9p8wrfxhpq3gbd8jzr42t04\" not found, persistentvolumeclaims \"brigade-worker-01c9pj36p3d20mcykwt96pnxag\" not found, persistentvolumeclaims \"brigade-worker-01c9q1xhe63rte5v7xzwq178dw\" not found, persistentvolumeclaims \"brigade-worker-01c9sgk31mawsbf3n3j1e0hypt\" not found, persistentvolumeclaims \"brigade-worker-01c9sjxxvv70naqd36aqng4set\" not found, persistentvolumeclaims \"brigade-worker-01c9sk95h6setspvny5rdhpzex\" not found, persistentvolumeclaims \"brigade-worker-01c9skqfw4dgj9d2bzh91vm805\" not found, persistentvolumeclaims \"brigade-worker-01c9sktkzcym5jymt3qfzyrf8j\" not found, persistentvolumeclaims \"brigade-worker-01c9p8wrfxhpq3gbd8jzr42t04\" not found, persistentvolumeclaims \"brigade-worker-01c9pj36p3d20mcykwt96pnxag\" not found, persistentvolumeclaims \"brigade-worker-01c9p8wrfxhpq3gbd8jzr42t04\" not found, persistentvolumeclaims \"brigade-worker-01c9pj36p3d20mcykwt96pnxag\" not found, persistentvolumeclaims \"brigade-worker-01c9q1xhe63rte5v7xzwq178dw\" not found, persistentvolumeclaims \"brigade-worker-01c9sk95h6setspvny5rdhpzex\" not found, persistentvolumeclaims \"brigade-worker-01c9skqfw4dgj9d2bzh91vm805\" not found, persistentvolumeclaims \"brigade-worker-01c9sktkzcym5jymt3qfzyrf8j\" not found]" key=heptio-ark/qa-20180423-3 logSource="pkg/controller/backup_controller.go:273"
```

How can I resolve this to backup my whole cluster?

**Comment by ncdc** [comment]
Hi @smm190, I'm guessing this is on Azure? If so, is it an AKS cluster?

**Comment by ncdc** [comment]
Do those PVCs exist?

Can you give us the backup log? `ark backup logs NAME`

**Comment by ncdc** [comment]
@smm190 Ark lists all pods (in your case, it found 124). For each pod, if it has any PVCs, it backs them up immediately instead of waiting until it lists all PVCs. I'm wondering if this is what happened:

1. Ark lists pods & gets 124 pods
1. Simultaneously:
    1. Ark loops through all the pods, backing up each one & its associated PVCs
    1. Some of the pods and their associated PVCs are deleted

Is that possible - could some of the pods & PVCs be deleted while Ark is backing up?

I also noticed some other errors in your log file where it failed to create disk snapshots because of an authentication issue. Did you specify your AKS cluster's `MC_....` resource group in the Ark cloud credentials file, or the resource group in which you created the cluster?

**Comment by smm190** [comment]
@ncdc It is not possible that some of the pods and PVCs were being deleted while Ark is backing up.  What is possible, however, is that this is an issue with Azure Brigade PVCs.  I notice all of the errors are for Brigade PVCs not being found.  Brigade is a tool for scripting job runs in Kubernetes.  It dynamically provisions PVCs when necessary for the jobs/pods, and it looks like those PVCs no longer exist.  I don't necessarily need those pods to be copied over, however.  They are just pods that ran once in a brigade job and are now complete (not currently doing anything).  If I want to spin them up again, I just re-run the brigade job.  Is there a way around this error, possibly by excluding the completed brigade pods? 

For the authentication issue you're seeing, I specified the resource group in which I created the cluster.  The documentation was a little bit confusing on this.  It wasn't clear to me what the "2nd" resource group it was referring to was.  My disks are provisioned in the same resource group as my virtual machines.  What is the "MC_" resource group you've mentioned?

**Comment by ncdc** [comment]
> They are just pods that ran once in a brigade job and are now complete (not currently doing anything)

Could you please paste or gist a sample of one of those pods? `kubectl get pod NAME -o yaml`.

> What is the "MC_" resource group you've mentioned?

When you create an AKS cluster, you create it in a preexisting resource group. Let's call it "foo" for this example. As part of the AKS cluster provisioning process, a new resource group is created, and it starts with "MC_" (let's call it "MC_bar"). This new resource group is used to hold everything related to your cluster - VMs, dynamically provisioned disks, load balancers, etc. You need to specify the new resource group (MC_bar) in the Ark cloud credentials file for snapshotting to work.

**Comment by smm190** [comment]
@ncdc Thanks for your quick replies.  I searched for the first missing PVC in my cluster and found the pod that used it.  I have attached the log from the pod where it states that it destroys the PVC upon completion.  
![successfulbrigadepod_pvcdestroyed](https://user-images.githubusercontent.com/35818471/39204946-e9670a96-47be-11e8-8856-9d600c0c98e2.PNG)

**Below is the yaml of the pod:**  
```
[ml2@ml2-centos-qa ark]$ kubectl get pod brigade-worker-01c9sm9g47trhpg0fmhwb53hc4 -o yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: 2018-03-29T19:38:28Z
  labels:
    build: 01c9sm9g47trhpg0fmhwb53hc4
    commit_id: ""
    component: build
    heritage: brigade
    project: brigade-830c16d4aaf6f5490937ad719afd8490a5bcbef064d397411043ac
  name: brigade-worker-01c9sm9g47trhpg0fmhwb53hc4
  namespace: default
  resourceVersion: "319685"
  selfLink: /api/v1/namespaces/default/pods/brigade-worker-01c9sm9g47trhpg0fmhwb53hc4
  uid: c12f2c1b-3388-11e8-a222-000d3a600c04
spec:
  containers:
  - command:
    - yarn
    - -s
    - start
    env:
    - name: CI
      value: "true"
    - name: BRIGADE_BUILD_ID
      value: 01c9sm9g47trhpg0fmhwb53hc4
    - name: BRIGADE_BUILD_NAME
      value: brigade-worker-01c9sm9g47trhpg0fmhwb53hc4
    - name: BRIGADE_COMMIT_ID
    - name: BRIGADE_COMMIT_REF
      value: master
    - name: BRIGADE_EVENT_PROVIDER
      value: brigade-cli
    - name: BRIGADE_EVENT_TYPE
      value: exec
    - name: BRIGADE_PROJECT_ID
      value: brigade-830c16d4aaf6f5490937ad719afd8490a5bcbef064d397411043ac
    - name: BRIGADE_REMOTE_URL
      value: https://github.com/deis/empty-testbed.git
    - name: BRIGADE_WORKSPACE
      value: /vcs
    - name: BRIGADE_PROJECT_NAMESPACE
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.namespace
    - name: BRIGADE_REPO_KEY
      valueFrom:
        secretKeyRef:
          key: sshKey
          name: brigade-830c16d4aaf6f5490937ad719afd8490a5bcbef064d397411043ac
          optional: true
    - name: BRIGADE_REPO_AUTH_TOKEN
      valueFrom:
        secretKeyRef:
          key: github.token
          name: brigade-830c16d4aaf6f5490937ad719afd8490a5bcbef064d397411043ac
          optional: true
    - name: BRIGADE_SERVICE_ACCOUNT
      value: brigade-worker
    image: deis/brigade-worker:v0.12.0
    imagePullPolicy: Always
    name: brigade-runner
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /etc/brigade
      name: brigade-build
      readOnly: true
    - mountPath: /vcs
      name: vcs-sidecar
      readOnly: true
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: brigade-worker-token-zs5gv
      readOnly: true
  dnsPolicy: ClusterFirst
  nodeName: k8s-lxagntd4-17778578-1
  nodeSelector:
    beta.kubernetes.io/os: linux
  restartPolicy: Never
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: brigade-worker
  serviceAccountName: brigade-worker
  terminationGracePeriodSeconds: 30
  volumes:
  - name: brigade-build
    secret:
      defaultMode: 420
      secretName: brigade-worker-01c9sm9g47trhpg0fmhwb53hc4
  - emptyDir: {}
    name: vcs-sidecar
  - name: brigade-worker-token-zs5gv
    secret:
      defaultMode: 420
      secretName: brigade-worker-token-zs5gv
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: 2018-03-29T19:38:28Z
    reason: PodCompleted
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: 2018-03-29T19:39:28Z
    reason: PodCompleted
    status: "False"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: 2018-03-29T19:38:28Z
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://f3b967e89523a03ec8d7bd883c328a4d91f037d91483d82d7ff5ff62a0774a38
    image: deis/brigade-worker:v0.12.0
    imageID: docker-pullable://deis/brigade-worker@sha256:f937b4375c3cd122b5fc7e365c06b4772cb5b43ed31c2065f28993ff246f2fe9
    lastState: {}
    name: brigade-runner
    ready: false
    restartCount: 0
    state:
      terminated:
        containerID: docker://f3b967e89523a03ec8d7bd883c328a4d91f037d91483d82d7ff5ff62a0774a38
        exitCode: 0
        finishedAt: 2018-03-29T19:39:28Z
        reason: Completed
        startedAt: 2018-03-29T19:38:30Z
  hostIP: 10.1.74.4
  phase: Succeeded
  podIP: 10.1.74.31
  qosClass: BestEffort
  startTime: `2018-03-29T19:38:28Z`
```


I was mistaken before.  The cluster is an acs-engine cluster - I thought that was synonymous with AKS.   That is why we don't have the MC_ resource group.  I provisioned the cluster myself into the resource group that I specified in Ark.  The resource  group I specified in the Ark configuration is the one that contains all of my cluster resources - I am only aware of this one resource group and I see all of the virtual machines & disks in it.  I see the bearer token refresh errors in the logs:

`time="2018-04-24T14:05:04Z" level=error msg="error creating snapshot" backup=heptio-ark/qa-20180424-1 error="rpc error: code = Unknown desc = azure.BearerAuthorizer#WithAuthorization: Failed to refresh the Token for request to https://management.azure.com/subscriptions/378079de-70e6-423f-bf1a-93947a02ee38/resourceGroups/metaverse-acs-k8s-qa/providers/Microsoft.Compute/disks/kubemasterqa-dynamic-pvc-9adeb07d-31ba-11e8-a222-000d3a600c04?api-version=2016-04-30-preview: StatusCode=0 -- Original Error: adal: Refresh request failed. Status Code = '401'" group=v1 groupResource=pods logSource="pkg/backup/item_backupper.go:322" name=pvc-9adeb07d-31ba-11e8-a222-000d3a600c04 namespace=default volumeID=kubemasterqa-dynamic-pvc-9adeb07d-31ba-11e8-a222-000d3a600c04`

Is the only cause of this error the mis-specification of the resource group?  I don't think another resource group exists.

**Comment by ncdc** [comment]
I filed #454 so we can avoid trying to restore completed pods and jobs. That's why you're getting the "PVC not found" errors.

Thanks for the clarification on ACS vs AKS. You can ignore my comments about a 2nd resource group, in this case. It looks like you're getting a "401 unauthorized" error when it's trying to talk to Azure. @skriss can you assist with debugging this one?

**Comment by skriss** [comment]
@smm190 did you use our example for setting up the service principal for Ark? Is it in the `Contributor` role for the subscription?

**Comment by skriss** [comment]
Usually when we see this error, it implies one of the fields in the `cloud-credentials` secret is incorrect. Could you double-check all of those to ensure they're correct? The volume-snapshot code uses a different subset of the secret's fields than the blob storage upload code so it could make sense that only snapshots are failing.

**Comment by skriss** [comment]
Probably - try running `kubectl -n heptio-ark scale deploy/ark --replicas 0` followed by `kubectl -n heptio-ark scale deploy/ark --replicas 1`

**Comment by ncdc** [comment]
You'll want to scale the ark deployment down to 0 and back to 1 to get it to pick up the updated secret, iirc.

**Comment by skriss** [comment]
jinx :)

**Comment by smm190** [comment]
@skriss Awesome.  After scaling the ark deployment down to 0 and back to 1, ark picked up the updated credentials and the snapshot authentication errors have disappeared from my backup log.  The only errors still left are the persistentvolumeclaim not found errors.  Thank you!

**Comment by skriss** [comment]
This has been resolved with #463, closing.
